contracts/market/cra/MarketLib.sol

```diff
@@ -0,0 +1,172 @@
+pragma solidity ^0.8.0;
+
+// external contracts
+import { ECDSA } from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+// Teller contracts
+import { PausableMods } from "../../settings/pausable/PausableMods.sol";
+import {
+    ReentryMods
+} from "../../contexts2/access-control/reentry/ReentryMods.sol";
+import { RolesMods } from "../../contexts2/access-control/roles/RolesMods.sol";
+import { ADMIN } from "../../shared/roles.sol";
+import {
+    Signature,
+    MarketStorageLib,
+    MarketStorage,
+    MarketConfig,
+    ProviderConfig,
+    LoanRequest
+} from "../../storage/market.sol";
+import { Verifier } from "./verifier.sol";
+import { LibLoans } from "../libraries/LibLoans.sol";
+// import reentry guard
+import "hardhat/console.sol";
+
+library MarketLib {
+    function s() internal pure returns (MarketStorage storage s_) {
+        s_ = MarketStorageLib.store();
+    }
+
+    function m(bytes32 marketId)
+        internal
+        view
+        returns (MarketConfig storage m_)
+    {
+        m_ = s().markets[marketId];
+    }
+
+    function p(bytes32 providerId)
+        internal
+        view
+        returns (ProviderConfig storage p_)
+    {
+        p_ = s().providers[providerId];
+    }
+
+    function c(bytes32 commitmentId) internal view returns (bool c_) {
+        c_ = s().usedCommitments[commitmentId];
+    }
+
+    function setProviderAdmin(
+        bytes32 providerId,
+        address account,
+        bool admin
+    ) internal onlyProviderAdmin(providerId) {
+        p(providerId).signer[account] = admin;
+    }
+
+    function getProviderAdmin(bytes32 providerId)
+        internal
+        view
+        returns (bool admin)
+    {
+        admin = p(providerId).signer[msg.sender];
+    }
+
+    function setProviderSigner(
+        bytes32 providerId,
+        address account,
+        bool signerValue
+    ) internal onlyProviderAdmin(providerId) {
+        p(providerId).signer[account] = signerValue;
+    }
+
+    function setProviderMaxAge(bytes32 providerId, uint32 maxAge)
+        internal
+        onlyProviderAdmin(providerId)
+    {
+        p(providerId).maxAge = maxAge;
+    }
+
+    function setMarketAdmin(
+        bytes32 marketId,
+        address account,
+        bool admin
+    ) internal onlyMarketAdmin(marketId) {
+        m(marketId).admin[account] = admin;
+    }
+
+    /**
+     * @notice it gets the user score and user request then returns the loan interest rate,
+     * loan collateral ratio and loan amount.
+     * @param marketScore the score of the user
+     * @param request the user's request for the loan
+     * @return userInterestRate returns the interest rate for the user based on his score
+     * @return userCollateralRatio returns the collateral ratio of the user based on his score
+     * @return userLoanAmount returns the amount for the user to take a loan out based on his score
+     */
+    function handler(uint256 marketScore, LoanRequest memory request)
```

Comment:
we will want this to be handled by a contract instead of a library so that we can have one deployed per market. when the transaction executes, we will use the `marketID` to look up a mapping that stores the contract address for the market and call it to return this data